# GenePS
Gene Positioning System

[![Build Status](https://travis-ci.org/jgraveme/GenePS.svg?branch=master)](https://travis-ci.org/jgraveme/GenePS)

UNDER CONSTRUCTION  !! :)

GenePS aims to find genes in draft genomes by exploiting the software HMMER, BLAST and Exonerate. The program is written
in Python 3!

SHORT MANUAL

Download all files of GenePS from Github. Most of them function merely as libraries but need to be present at
the current stage. Furthermore the following programs and python packages are required to run GenePS:
Programs:
- HMMER
- MAFFT
- BLAST
- Exonerate
- Trimal

Python package:
- (Python 3)
- docopt
- matplotlib.pyplot
- seaborn

GenePS consists of two separate scripts: "make_GenePS" processing the input and "run_GenePS" which does the
actual gene finding.


1) make_GenePS.py
~~~
Usage: make_GenePS.py                     -i <DIR> -o <DIR> [-b <DIR>] [-t <FILE>] [-p <FILE>]

    -i, --input <DIR>                     either single fasta-file or a directory with files and/or sub-folders
    -o, --output <DIR>                    creates a directory to store the output

    optional arguments (in case additional data from orthofinder is available)
    -b, --blast <DIR>                     directory containing within-species-BLAST text files (one per species)
    -t, --translation_file <FILE>         file to translate species/protein names to ids (69_25600: MJAVA.g25693.t1)
    -p, --protein_fasta <FILE>            fasta-file providing all protein sequences with name as header
~~~
The input can be either a single fasta file containing PROTEIN sequences from one gene family but different species, or
a whole directory with sub-folders. This allows to gather many gene families or files respectively under one group name.

The output will be exactly one file per "group" holding information required for the next run_GenePS.py step.
This "makeGenePS-file" will be stored in the specified output directory, together with gene family specific HMM files.


2) run_GenePS.py
~~~
Usage: run_GenePS.py                          -m <DIR> -g <FILE> -c <INT> -f <INT>

        General
        -m, --GenePS_result_dir <DIR>         output directory from make_GenePS.py
        -g, --genome <FILE>                   Target genome

        optional arguments
        -c, --coverage_filer <INT>            Minimum length in percent, with which a Blast query should align to the target genome (used to filter Blast hits before passing them to exonerate)
        -f, --HMM_filter <INT>                Factor to multiply standard deviation of the HMM score distribution (validate predictions)
~~~

Give run_GenePS the output directory generated by make_GenePS. It will then search for the file suffix and load all
makeGenePS files in the specified folder. You also need to tell GenePS the path to the target genome. The genome should be
provided as a fasta file. GenePS takes care of generating a genome Blast database.
The Prediction will then be printed to a the screen in style of:

Gene, family name,  Contig,  start,  end,  status (whether it passed all filtering steps)

#### Example

[+] eef_all_proteins_of_proteom_test, I, 9158949, 9171695                PASSED

For all passed and filtered Predictions of a gene family, a fasta-file as wel as a gff-file will be generated and saved to a newly created "Prediction"-folder within the original output directory.
